File: app/src/main/java/eu/kliq/moviequiz/MainActivity.java
package eu.kliq.moviequiz;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PorterDuff;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

import info.movito.themoviedbapi.TmdbApi;
import info.movito.themoviedbapi.TmdbMovies;
import info.movito.themoviedbapi.model.Artwork;
import info.movito.themoviedbapi.model.MovieDb;
import info.movito.themoviedbapi.model.MovieImages;
import info.movito.themoviedbapi.model.config.TmdbConfiguration;
import info.movito.themoviedbapi.model.core.MovieResultsPage;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private static final String TAG = MainActivity.class.getSimpleName();
    private static final String API_KEY = BuildConfig.TMDB_API_KEY;
    private static final String BACKDROP_SIZE = "w1280";
    private static final int CORRECT_ANWSER_COLOR = 0xFF00FF00;
    private static final int WRONG_ANWSER_COLOR = 0xFFFF0000;

    // TheMovieDB API objects
    TmdbConfiguration mTmdbConfiguration;
    TmdbMovies mTmdbMovies;

    boolean isWaiting = true;
    int mCurrentPage = 0;
    int mCurrentQuestion = 0;
    int mCurrentMovieId;
    Bitmap mCurrentImage = null;
    Map<Integer, Integer> mCurrentMovies = new HashMap<>();
    int mCorrectAnswer;

    String mBaseUrl;
    Map<Integer, MovieDb> mMovies;
    Random mRandom = new Random();

    // UI
    ImageView mImageView;
    List<Button> mButtons;
    private RelativeLayout mWaitingLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mWaitingLayout = (RelativeLayout)findViewById(R.id.waiting_layout);
        mImageView = (ImageView)findViewById(R.id.movie_img);
        mButtons = new ArrayList<>();
        mButtons.add((Button)findViewById(R.id.answer_button1));
        mButtons.add((Button)findViewById(R.id.answer_button2));
        mButtons.add((Button)findViewById(R.id.answer_button3));
        mButtons.add((Button)findViewById(R.id.answer_button4));

        for (Button button : mButtons) {
            button.setOnClickListener(this);
        }

        ConnectivityManager connMgr = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
        if (networkInfo != null && networkInfo.isConnected()) {
            initialize();
        }
    }

    private void initialize() {
        new GetPopularMoviesTask().execute();
    }

    private void generateMoviesMap(List<MovieDb> results) {
        mMovies = new LinkedHashMap<>();
        for (MovieDb movie : results) {
            int id = movie.getId();
            mMovies.put(id, movie);
        }
        generateQuestion();
    }

    private void generateQuestion() {
        setWaitingState(true);
        mCurrentMovies.clear();
        mCurrentMovieId = getRandomMovieId();
        getMovieImage(mCurrentMovieId);
        mCorrectAnswer = mRandom.nextInt(3);
        int id;
        for (int iteration = 0; iteration < 4; iteration++) {
            // If this iteration is the same as the correct anwser, set the ID of the current movie
            if (iteration == mCorrectAnswer) {
                id = mCurrentMovieId;
            } else {
                id = getUniqueRandomMovieId();
            }
            mCurrentMovies.put(iteration, id);
        }
        mCurrentQuestion++;
    }

    private int getUniqueRandomMovieId() {
        int id;
        do {
            id = getRandomMovieId();
        } while (mCurrentMovies.containsValue(id));
        return id;
    }

    private int getRandomMovieId() {
        final int randomEntry = mRandom.nextInt(mMovies.size() - 1);
        return (int) mMovies.keySet().toArray()[randomEntry];
    }

    private void setUiForCurrentQuestion() {
        Log.d(TAG, "setUiForCurrentQuestion()");
        // Set image
        mImageView.setImageBitmap(mCurrentImage);
        // Set buttons
        for (int iteration = 0; iteration < 4; iteration++) {
            final int id = mCurrentMovies.get(iteration);
            final String title = mMovies.get(id).getTitle();
            final Button button = mButtons.get(iteration);
            button.setText(title);
            button.getBackground().setColorFilter(null);
        }
        setWaitingState(false);
    }

    private void getMovieImage(int id) {
        Log.d(TAG, "getMovieImage() id: " + id);
        new GetMovieImageTask().execute(id);
    }

    @Override
    public void onClick(View v) {
        if (!isWaiting) {
            switch (v.getId()) {
                case R.id.answer_button1:
                    selectAnwser(0);
                    break;
                case R.id.answer_button2:
                    selectAnwser(1);
                    break;
                case R.id.answer_button3:
                    selectAnwser(2);
                    break;
                case R.id.answer_button4:
                    selectAnwser(3);
                    break;
            }
        }
    }

    private void selectAnwser(int anwserId) {
        Log.d(TAG, "selectAnwser() anwserId: " + anwserId);
        final int color;
        if (anwserId == mCorrectAnswer) {
            Log.d(TAG, "selectAnwser() true");
            color = CORRECT_ANWSER_COLOR;
        } else {
            Log.d(TAG, "selectAnwser() false");
            color = WRONG_ANWSER_COLOR;
        }
        mButtons.get(anwserId).getBackground().setColorFilter(color, PorterDuff.Mode.MULTIPLY);
        generateQuestion();
    }

    private void setWaitingState(boolean state) {
        if (state) {
            mWaitingLayout.setVisibility(View.VISIBLE);
        } else {
            mWaitingLayout.setVisibility(View.INVISIBLE);
        }
        isWaiting = state;
    }

    private class GetPopularMoviesTask extends AsyncTask<Void, Void, MovieResultsPage> {
        @Override
        protected MovieResultsPage doInBackground(Void... params) {
            mTmdbMovies = new TmdbApi(API_KEY).getMovies();
            mTmdbConfiguration = new TmdbApi(API_KEY).getConfiguration();
            return mTmdbMovies.getPopularMovieList("pl", mCurrentPage++);
        }

        @Override
        protected void onPostExecute(MovieResultsPage page) {
            mBaseUrl = mTmdbConfiguration.getBaseUrl();
            generateMoviesMap(page.getResults());
        }
    }

    private class GetMovieImageTask extends AsyncTask<Integer, Void, Bitmap> {
        @Override
        protected Bitmap doInBackground(Integer... params) {
            Bitmap image = null;
            List<String> paths = new ArrayList<>();
            MovieImages movieImages = mTmdbMovies.getImages(params[0], "");
            List<Artwork> backdrops = movieImages.getBackdrops();
            for (Artwork backdrop : backdrops) {
                if (backdrop.getLanguage() == null) {
                    paths.add(backdrop.getFilePath());
                }
            }
            int randomPathId = mRandom.nextInt(paths.size() - 1);
            try {
                URL url = new URL(mBaseUrl + BACKDROP_SIZE + paths.get(randomPathId));
                image = BitmapFactory.decodeStream(url.openConnection().getInputStream());
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
            return image;
        }

        @Override
        protected void onPostExecute(Bitmap result) {
            mCurrentImage = result;
            setUiForCurrentQuestion();
        }
    }
}

File: app/src/main/java/eu/kliq/moviequiz/MainActivity.kt
package eu.kliq.moviequiz

import android.content.Context
import android.graphics.Bitmap
import android.graphics.PorterDuff
import android.net.ConnectivityManager
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"
    private var isWaiting = true
    private lateinit var mButtons: MutableList<Button>
    private val CORRECT_ANSWER_COLOR = 0xFF00FF00.toInt()
    private val WRONG_ANSWER_COLOR = 0xFFFF0000.toInt()
    private val mQuestionManager: QuestionManager = QuestionManager
    private var startTime: Long = 0
    private var mHiScore = 0
    private val HI_SCORE_KEY = "hi_score_key"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        addButtons()
        refreshScoreText()
        getHiScore()
        overlay_layout.setOnClickListener({restartGame()})
        checkConnectionAndInit()
    }

    private fun addButtons() {
        mButtons = ArrayList()
        for (iteration in 0 until QuestionManager.BUTTONS) {
            val button = Button(this)
            button.setOnClickListener({
                if (!isWaiting) selectAnswer(iteration)
            })
            button_layout.addView(button)
            mButtons.add(button)
        }
    }

    private fun checkConnectionAndInit() {
        val connMgr = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = connMgr.activeNetworkInfo
        if (networkInfo != null && networkInfo.isConnected) {
            mQuestionManager.initialize {generateQuestion()}
        }
    }

    private fun setUiForCurrentQuestion(bitmap: Bitmap) {
        Log.d(TAG, "setUiForCurrentQuestion()")
        // Set image
        movie_img.setImageBitmap(bitmap)
        // Set buttons
        for (iteration in 0 until QuestionManager.BUTTONS) {
            val button = mButtons[iteration]
            button.text = mQuestionManager.getCurrentQuestionMovieTitle(iteration)
            button.background.colorFilter = null
        }
        startTime =  System.currentTimeMillis()
        refreshRoundText()
        setWaitingState(false)
    }

    private fun selectAnswer(answerId: Int) {
        Log.d(TAG, "selectAnswer($answerId)")
        val time = System.currentTimeMillis() - startTime
        val isCorrect = mQuestionManager.onAnswer(answerId, time)
        val color = if (isCorrect) CORRECT_ANSWER_COLOR else WRONG_ANSWER_COLOR
        mButtons[answerId].background.setColorFilter(color, PorterDuff.Mode.MULTIPLY)
        refreshScoreText()
        if (isCorrect) generateQuestion() else finishGame()
    }

    private fun finishGame() {
        overlay_layout.visibility = View.VISIBLE
        game_over_layout.visibility = View.VISIBLE
        saveHiScore()
    }

    private fun restartGame() {
        overlay_layout.visibility = View.GONE
        game_over_layout.visibility = View.GONE
        mQuestionManager.restartGame()
        refreshScoreText()
        generateQuestion()
    }

    private fun refreshScoreText() {
        score_text.text = getString(R.string.score, mQuestionManager.score)
    }

    private fun refreshRoundText() {
        round_text.text = getString(R.string.round, mQuestionManager.currentQuestion)
    }

    private fun refreshHiScoreText() {
        hi_score_text.text = getString(R.string.hi_score, mHiScore)
    }

    private fun generateQuestion() {
        setWaitingState(true)
        mQuestionManager.generateQuestion { bitmap -> setUiForCurrentQuestion(bitmap) }
    }

    private fun setWaitingState(state: Boolean) {
        overlay_layout.visibility = if (state) View.VISIBLE else View.GONE
        waiting_progressbar.visibility = if (state) View.VISIBLE else View.GONE
        isWaiting = state
    }

    private fun getHiScore() {
        mHiScore = getPreferences(Context.MODE_PRIVATE).getInt(HI_SCORE_KEY, 0)
        refreshHiScoreText()
    }

    private fun saveHiScore() {
        if (mQuestionManager.score > mHiScore) {
            mHiScore = mQuestionManager.score
            val sharedPref = getPreferences(Context.MODE_PRIVATE)
            val editor = sharedPref.edit()
            editor.putInt(HI_SCORE_KEY, mHiScore)
            editor.apply()
            refreshHiScoreText()
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        mQuestionManager.restartGame()
    }
}
